packages/babel-traverse/src/path/lib/virtual-types.ts

```diff
@@ -178,7 +111,5 @@ export const NumericLiteralTypeAnnotation: Wrapper = {
 
 export const ForAwaitStatement: Wrapper = {
   types: ["ForOfStatement"],
-  checkPath({ node }: NodePath<t.ForOfStatement>): boolean {
-    return node.await === true;
-  },
+  checkPath: path => path.isForAwaitStatement(),
```

Comment:
Can we simplify `Wrapper` to remove `checkPath`? (I'm not sure if this is possible, I try to understand the codes but they are a bit complicated)

Reply:
Nope. The `checkPath` applies additional selection logic based on `Wrapper#types`.

https://github.com/babel/babel/blob/029fa1778296ad2e854313d2582aacf3a0d74d43/packages/babel-traverse/src/visitors.ts#L285-L287

Say if we have a `ReferencedIdentifier` visitor, Babel knows an `Identifier` could be a `ReferencedIdentifier` from `Wrapper#types`. Then it wraps the visitor so that the wrapped visitor will first run `wrapper.checkPath`, if the node is a `ReferencedIdentifier`, the wrapped visitor then applies the original visitor to the `Identifier` node, otherwise the visitor is inactivated.

Reply:
Could we pass `nodeType` to https://github.com/babel/babel/blob/029fa1778296ad2e854313d2582aacf3a0d74d43/packages/babel-traverse/src/visitors.ts#L66, and then rewrite the `wrapCheck` function so that it doesn't rely anymore on `checkPath`? Every `checkPath` function just forwards the check to the `.is*` method with the same name as the virtual type.
```ts
function wrapCheck(nodeType: String, fn: Function) {
  const newFn = function (this: unknown, path: NodePath) {
    if (path[`is${nodeType}`]()) {
      return fn.apply(this, arguments);
    }
  };
  newFn.toString = () => fn.toString();
  return newFn;
}
```

Or, we can make `checkPath` just the check method name (as a string), to avoid the unnecessary intermediate function:
```js
import * as virtualValidators from "./virtual-types-validator.ts"

export const ForAwaitStatement: Wrapper = {
  types: ["ForOfStatement"],
  checkPath: virtualValidators.isForAwaitStatement,
};
```
```js
function wrapCheck(wrapper: Wrapper, fn: Function) {
  const newFn = function (this: unknown, path: NodePath) {
    if (wrapper.checkPath.call(path)) {
      return fn.apply(this, arguments);
    }
  };
  newFn.toString = () => fn.toString();
  return newFn;
}
```